website/src/documentation/io/developing-io-java.md

```diff
@@ -0,0 +1,419 @@
+total data set.
+
+If you implement `splitAtFraction`, you must implement both `splitAtFraction`
+and `getFractionConsumed` in a thread-safe manner, or data loss is possible. You
+should also unit-test your implementation exhaustively to avoid data duplication
+or data loss.
+
+To ensure that your code is thread-safe, use the `RangeTracker` thread-safe
+helper object to manage positions in your data source when implementing
+`splitAtFraction` and `getFractionConsumed`.
+
+We highly recommended that you unit test your implementations of
+`splitAtFraction` using the `SourceTestUtils` class. `SourceTestUtils` contains
+a number of methods for testing your implementation of `splitAtFraction`,
+including exhaustive automatic testing.
+
+### Implementing wrapper transforms
+
+When you create a source that end-users will use, we recommend that
+you do not expose the code for the source itself. Instead, create a wrapping
+`PTransform`. See [PTransform wrappers](#ptransform-wrappers) to see how
+and why to avoid exposing your sources.
+
+### Convenience Source and Reader base classes
+
+The Beam SDK contains some convenient abstract base classes to help you create
+`Source` and `Reader` classes that work with common data storage formats, like
+files.
+
+#### FileBasedSource
+
+If your data source uses files, you can derive your `Source` and `Reader`
+classes from the `FileBasedSource` and `FileBasedReader` abstract base classes.
+`FileBasedSource` is a bounded source subclass that implements code common to
+Beam sources that interact with files, including:
+
+  * File pattern expansion
+  * Sequential record reading
+  * Split points
+
+
+## Implementing the FileBasedSink interface
+
+If your data source uses files, you can use the `FileBasedSink` interface to
+create a file-based sink for your pipeline. If your source is not file-based,
+implement a composite `PTransform` that uses `ParDo`  as discussed in the
+[developing I/O connectors overview]({{ site.baseurl }}/documentation/io/developing-io-overview/)
+
+When using the `FileBasedSink` interface, you must provide the format-specific
+logic that tells the runner how to write bounded data from your pipeline's
+`PCollection`s to an output sink. The runner writes bundles of data in parallel
+using multiple workers.
+
+Supply the logic for your file-based sink by implementing the following classes:
+
+  * A subclass of the abstract base class `FileBasedSink`. `FileBasedSink`
+    describes a location or resource that your pipeline can write to in
+    parallel.
+
+  * A user-facing wrapper `PTransform` that, as part of the logic, calls
+    [WriteFiles](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java)
+    and passes your `FileBasedSink` as a parameter.
+
+### Implementing the FileBasedSink subclass
+
+You must create a subclass of the abstract base class `FileBasedSink`. To avoid
```

Comment:
To avoid exposing your sink to end-users, your `FileBasedSink` sub-class should be protected or private.